### TCGA prognostic associations {.tabset}

* Based on data from the [Human Protein Atlas - Pathology Atlas](https://www.proteinatlas.org/humanproteome/pathology), we are here listing significant results from correlation analyses of mRNA expression levels of human genes in tumor tissue and the clinical outcome (survival) for ~8,000 cancer patients (TCGA)
* All correlation analyses have been performed in a gene-centric manner, and associations are only shown for genes in the target set. We separate between 
  * <span style="background-color:#005a32; padding:2x; color:white"><b>&nbsp; Favorable associations &nbsp; </b></span> : high expression of a given gene is associated with better survival
  * <span style="background-color:#99000d; padding:2x; color:white"><b>&nbsp; Unfavorable associations &nbsp; </b></span> : high expression of a given gene is associated with worse survival
* Strength of associations are provided through p-values (only associations with a p-value< = 0.001 are provided), in addition we provide a percentile rank for associations considering 
  a) all significant (p-value <= 0.001) associations across all tumor sites, and 
  b) only significant (p-value <= 0.001) associations found in the same tumor site


<br><br>

```{r filter_hpa_prognostic, echo=F, results='asis'}

show_hpa_prognostic <- list()
missing_hpa_prognostic <- list()
prognostic_assoc_data <- list()

for(n in c('favorable','unfavorable')){
    show_hpa_prognostic[[n]] <- TRUE
    missing_hpa_prognostic[[n]] <- FALSE
    prognostic_assoc_data[[n]] <- data.frame()
  
    prognostic_assoc_data[[n]] <- 
      onc_enrich_report[['data']][['cancer_prognosis']][['assocs']] %>% 
      dplyr::filter(evidence_direction == n) %>%
      dplyr::select(-evidence_direction)
    
    if(nrow(prognostic_assoc_data[[n]]) == 0){
      missing_hpa_prognostic[[n]] <- TRUE
      show_hpa_prognostic[[n]] <- FALSE
    }
}

```


#### Favorable associations

```{r table_browse_hpa_prognostic_favorable, echo=F, results = "asis", eval = show_hpa_prognostic[['favorable']]}
library(crosstalk)


prognostic_genes_favorable <- crosstalk::SharedData$new(prognostic_assoc_data[['favorable']])

crosstalk::bscols(
  list(
    crosstalk::filter_select("symbol", "Target gene", prognostic_genes_favorable, ~symbol),
    crosstalk::filter_slider("log10_p_value", "-log10 (p-value)", 
                             prognostic_genes_favorable, ~log10_p_value, step = 1,
                             min = 3, max = 16)
  ),
  list(
    crosstalk::filter_select("primary_site", "Primary tumor site", prognostic_genes_favorable, ~primary_site),
    crosstalk::filter_slider("percentile_rank_all", "Percentile rank (any tumor site)", 
                             prognostic_genes_favorable, ~percentile_rank_all, 
                             step = 10, min = 0, max = 100),
    crosstalk::filter_slider("percentile_rank_site", "Percentile rank (tumor site)", 
                             prognostic_genes_favorable, ~percentile_rank_site, 
                             step = 10, min = 0, max = 100)

  )
)

htmltools::br()
DT::datatable(prognostic_genes_favorable, escape = F, 
              extensions=c("Buttons","Responsive"), width = "100%",
              options=list(buttons = c('csv','excel'), 
                           dom = 'Bfrtip')) %>%
  DT::formatStyle("symbol","log10_p_value", 
                  color = "white", 
                  fontWeight = "bold",
                  backgroundColor = 
                    DT::styleInterval(onc_enrich_report[['config']][['prognosis']][['breaks']], onc_enrich_report[['config']][['prognosis']][['colors_favorable']]))


```


```{r prognostic_missing_favorable, echo=F, results = 'asis', eval = missing_hpa_prognostic[['favorable']]}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;No genes in target set with favorable gene expression association to prognosis&nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```

<br><br><br>


#### Unfavorable associations

```{r table_browse_hpa_prognostic_unfavorable, echo=F, results = "asis", eval = show_hpa_prognostic[['unfavorable']]}
library(crosstalk)


prognostic_genes_unfavorable <- crosstalk::SharedData$new(prognostic_assoc_data[['unfavorable']])

crosstalk::bscols(
  list(
    crosstalk::filter_select("symbol", "Target gene", prognostic_genes_unfavorable, ~symbol),
    crosstalk::filter_slider("log10_p_value", "-log10 (p-value)", 
                             prognostic_genes_unfavorable, ~log10_p_value, step = 1,
                             min = 3, max = 16)
  ),
  list(
    crosstalk::filter_select("primary_site", "Primary tumor site", prognostic_genes_unfavorable, ~primary_site),
    crosstalk::filter_slider("percentile_rank_all", "Percentile rank (any tumor site)", 
                             prognostic_genes_unfavorable, ~percentile_rank_all, 
                             step = 10, min = 0, max = 100),
    crosstalk::filter_slider("percentile_rank_site", "Percentile rank (tumor site)", 
                             prognostic_genes_unfavorable, ~percentile_rank_site, 
                             step = 10, min = 0, max = 100)

  )
)

htmltools::br()
DT::datatable(prognostic_genes_unfavorable, escape = F, 
              extensions=c("Buttons","Responsive"), width = "100%",
              options=list(buttons = c('csv','excel'), 
                           dom = 'Bfrtip')) %>%
  DT::formatStyle("symbol","log10_p_value", 
                  color = "white", 
                  fontWeight = "bold",
                  backgroundColor = 
                    DT::styleInterval(onc_enrich_report[['config']][['prognosis']][['breaks']], onc_enrich_report[['config']][['prognosis']][['colors_unfavorable']]))


```
<br><br>

```{r prognostic_missing_unfavorable, echo=F, results = 'asis', eval = missing_hpa_prognostic[['unfavorable']]}
cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;No genes in target set with unfavorable gene expression association to prognosis&nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')
```

<br><br><br>
