### TCGA aberration frequency

* The somatic mutation frequencies of the target genes in ~10,000 tumor samples have been calculated with data from [The Cancer Genome Atlas (TCGA)](https://portal.gdc.cancer.gov/)

<br><br>

#### SNVs/InDels {.tabset}

* Somatic SNVs/InDels in the target genes (top mutated) are illustrated with [oncoplots](https://www.bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/oncoplots.html)
* Gene mutation frequencies are sorted by type of diagnosis (i.e. cancer subtypes)

<br><br>

```{r tcga_oncoplot_setup, echo = F, eval = T}

tcga_oncoplot_data <- onc_enrich_report[["data"]][["tcga"]][["aberration"]][["table"]][["snv_indel"]]

gene_mar <- 8
annotationFontSize <- 1.8
sepwd_genes <- 0.7
fontSize <- 1.6
legendFontSize <- 2
legend_height <- 6
anno_height <- 1.2
fheight <- 21
fwidth <- 18

```


##### Breast

```{r tcga_oncoplot_breast, echo = F, eval = NROW(tcga_oncoplot_data[["Breast"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Breast"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Breast"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F, 
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_breast_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Breast"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Colon/Rectum

```{r tcga_oncoplot_colon, echo = F, eval = NROW(tcga_oncoplot_data[["Colon/Rectum"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Colon/Rectum"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Colon/Rectum"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis","MSI_status"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_colon_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Colon/Rectum"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Lung

```{r tcga_oncoplot_lung, echo = F, eval = NROW(tcga_oncoplot_data[["Lung"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Lung"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Lung"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_lung_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Lung"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Skin

```{r tcga_oncoplot_skin, echo = F, eval = NROW(tcga_oncoplot_data[["Skin"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Skin"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Skin"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_skin_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Skin"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Esophagus/Stomach

```{r tcga_oncoplot_stomach, echo = F, eval = NROW(tcga_oncoplot_data[["Esophagus/Stomach"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Esophagus/Stomach"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Esophagus/Stomach"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis","MSI_status"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_esophagus_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Esophagus/Stomach"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Cervix

```{r tcga_oncoplot_cervix, echo = F, eval = NROW(tcga_oncoplot_data[["Cervix"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Cervix"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Cervix"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_cervix_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Cervix"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Prostate

```{r tcga_oncoplot_prostate, echo = F, eval = NROW(tcga_oncoplot_data[["Prostate"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Prostate"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Prostate"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_prostate_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Prostate"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Ovary/Fallopian Tube

```{r tcga_oncoplot_ovary, echo = F, eval = NROW(tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_ovary_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Ovary/Fallopian Tube"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Uterus


```{r tcga_oncoplot_uterus, echo = F, eval = NROW(tcga_oncoplot_data[["Uterus"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Uterus"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Uterus"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis","MSI_status"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_uterus_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Uterus"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Pancreas


```{r tcga_oncoplot_pancreas, echo = F, eval = NROW(tcga_oncoplot_data[["Pancreas"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Pancreas"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Pancreas"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_pancreas_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Pancreas"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Soft Tissue


```{r tcga_oncoplot_softtissue, echo = F, eval = NROW(tcga_oncoplot_data[["Soft Tissue"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Soft Tissue"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Soft Tissue"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_softtissue_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Soft Tissue"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Myeloid


```{r tcga_oncoplot_myeloid, echo = F, eval = NROW(tcga_oncoplot_data[["Myeloid"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Myeloid"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Myeloid"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_myeloid_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Myeloid"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### CNS/Brain


```{r tcga_oncoplot_cns, echo = F, eval = NROW(tcga_oncoplot_data[["CNS/Brain"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["CNS/Brain"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["CNS/Brain"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_brain_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["CNS/Brain"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Liver


```{r tcga_oncoplot_liver, echo = F, eval = NROW(tcga_oncoplot_data[["Liver"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Liver"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Liver"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_liver_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Liver"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Kidney


```{r tcga_oncoplot_kidney, echo = F, eval = NROW(tcga_oncoplot_data[["Kidney"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Kidney"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Kidney"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_kidney_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Kidney"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Lymphoid


```{r tcga_oncoplot_lymphoid, echo = F, eval = NROW(tcga_oncoplot_data[["Lymphoid"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Lymphoid"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Lymphoid"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_lymphoid_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Lymphoid"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Head and Neck


```{r tcga_oncoplot_headneck, echo = F, eval = NROW(tcga_oncoplot_data[["Head and Neck"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Head and Neck"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Head and Neck"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```


```{r tcga_headneck_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Head and Neck"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Biliary Tract


```{r tcga_oncoplot_biliarytract, echo = F, eval = NROW(tcga_oncoplot_data[["Biliary Tract"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Biliary Tract"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Biliary Tract"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_biliary_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Biliary Tract"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Bladder/Urinary Tract


```{r tcga_oncoplot_bladder, echo = F, eval = NROW(tcga_oncoplot_data[["Bladder/Urinary Tract"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Bladder/Urinary Tract"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Bladder/Urinary Tract"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_bladder_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Bladder/Urinary Tract"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Pleura


```{r tcga_pleura, echo = F, eval = NROW(tcga_oncoplot_data[["Pleura"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Pleura"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Pleura"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_pleura_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Pleura"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```


##### Thyroid


```{r tcga_thyroid, echo = F, eval = NROW(tcga_oncoplot_data[["Thyroid"]][['top_mutated_genes']]) >= 5, fig.height = fheight, fig.width = fwidth}

code <- tcga_oncoplot_data[["Thyroid"]][['code']]
maf <- readRDS(url(paste0("https://github.com/sigven/oncoEnrichR/raw/master/maf/",code,".maf.rds")))
maftools::oncoplot(maf, genes = tcga_oncoplot_data[["Thyroid"]][['top_mutated_genes']]$symbol,
                     drawRowBar = F, drawColBar = F,
                     sortByAnnotation = T,
                     clinicalFeatures = c("Diagnosis"), showTitle = F,
                   gene_mar = gene_mar,
                   annotationFontSize = annotationFontSize, 
                   sepwd_genes = sepwd_genes, 
                   fontSize = fontSize,
                   legendFontSize = legendFontSize, 
                   legend_height = legend_height,
                   anno_height = anno_height)
```

```{r tcga_thyroid_missing, results = 'asis', echo = F, eval = NROW(tcga_oncoplot_data[["Thyroid"]][['top_mutated_genes']]) < 5,}

cat('<br><br>\n <ul><li>  <i> <span style="font-size: 105%; padding: 3px; background-color:#989898; color:white">&nbsp;&nbsp;<b>Too few genes (n < 5) </b> with significant SNV/InDel mutation frequency for oncoplot &nbsp;&nbsp; </span></i></li></ul>',sep='\n')
cat('\n')
cat('<br><br>')

```

<br>

#### Copy number alterations

* Genes targeted by somatic copy number alterations (sCNAs) in tumor samples have been retrieved from TCGA, where copy number state have been estimated with [GISTIC](http://portals.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t)
* Gene aberration frequency are plotted across two categories of mutation types
    1. <span style="background-color:#005a32; color:white"><b>&nbsp; sCNA - amplifications &nbsp; </b></span>
    2. <span style="background-color:#800026; color:white"><b>&nbsp; sCNA - homozygous deletions &nbsp; </b></span>
* Frequency is plotted pr. primary site as 
  * *_percent_mutated_* (percent of all tumor samples with the gene amplified/lost)
  * genes in the plot are ranked according to alteration frequency across all sites (_pancancer_),
  limited to the top 75 genes in the target set
* Frequencies across all subtypes per primary site are listed in an interactive table


<br><br>
```{r tcga_ampl_plot, echo = F, eval = T}

suppressMessages(library(plotly))
gene_aberration_top_mat <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_ampl']]
plotly_colors <- "YlGn"


fig <- plotly::plot_ly(
  colors = plotly_colors,
  width = 800,
  height = 400 + (14.67 * NROW(onc_enrich_report$data$tcga$aberration$matrix$cna_ampl))) %>%
    plotly::add_heatmap(
      y = rownames(gene_aberration_top_mat),
      x = colnames(gene_aberration_top_mat),
      z = gene_aberration_top_mat,
      hovertext = "Percent mutated",
      yaxis = "y") %>%
    plotly::layout(
      title = "Amplifications",
      xaxis = list(tickfont = list(size = 13, family = "Helvetica"), 
                   tickangle = -50),
      yaxis = list(tickfont = list(size = 12, family = "Helvetica")),
      margin = list(l = 75, r = 20, b = 150, t = 30, pad = 4),
      legend = list(orientation = 'h')
    ) %>%
    plotly::colorbar(
      nticks = 10,
      title = list(text = "Percent mutated",side = "bottom"),
      limits = c(0, plyr::round_any(max(gene_aberration_top_mat), 10, ceiling)))
  fig


```
<br><br>
```{r tcga_homdel_plot, echo = F, eval = T}

suppressMessages(library(plotly))
gene_aberration_top_mat <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['matrix']][['cna_homdel']]

plotly_colors <- "YlOrRd"

fig <- plotly::plot_ly(
  colors = plotly_colors,
  width = 800,
  height = 400 + (14.67 * NROW(gene_aberration_top_mat))) %>%
    plotly::add_heatmap(
      y = rownames(gene_aberration_top_mat),
      x = colnames(gene_aberration_top_mat),
      z = gene_aberration_top_mat,
      hovertext = "Percent mutated",
      yaxis = "y") %>%
    plotly::layout(
      title = 'Homozygous deletions',
      xaxis = list(tickfont = list(size = 13, family = "Helvetica"), 
                   tickangle = -50),
      yaxis = list(tickfont = list(size = 12, family = "Helvetica")),
      margin = list(l = 75, r = 20, b = 150, t = 30, pad = 4)
    ) %>%
    plotly::colorbar(
      nticks = 10, 
      title = list(text = "Percent mutated",side = "bottom"),
      limits = c(0, plyr::round_any(max(gene_aberration_top_mat), 10, ceiling)))
  fig


```
<br><br>




```{r filter_tcga, echo=F, results='asis'}

show_tcga_freqs <- FALSE
missing_tcga_freqs <- TRUE
all_targets_vtypes <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['table']][['cna_homdel']] %>% 
  dplyr::bind_rows(
    onc_enrich_report[['data']][['tcga']][['aberration']][['table']][['cna_ampl']]
    )

if(nrow(all_targets_vtypes) > 0){
  show_tcga_freqs <- TRUE
  missing_tcga_freqs <- FALSE
}

```


```{r table_browse_tcga_freqs, echo=F, results = "asis", eval = show_tcga_freqs}
library(crosstalk)

all_targets_vtypes <- 
  onc_enrich_report[['data']][['tcga']][['aberration']][['table']][['cna_homdel']] %>% 
  dplyr::bind_rows(
    onc_enrich_report[['data']][['tcga']][['aberration']][['table']][['cna_ampl']]
    ) %>%
  dplyr::arrange(desc(percent_mutated))
targets_tcga <- crosstalk::SharedData$new(all_targets_vtypes)

crosstalk::bscols(
  list(
    crosstalk::filter_select("primary_site", "Primary site/tissue", targets_tcga, ~primary_site),
    crosstalk::filter_select("variant_type", "Aberration type", targets_tcga, ~variant_type),
    crosstalk::filter_slider("percent_mutated", "Percent mutated", 
                             targets_tcga, ~percent_mutated, step = 10,
                             min = 0, max = 100)
  ),
  list(
    crosstalk::filter_select("primary_diagnosis", "Diagnosis", targets_tcga, ~primary_diagnosis),
    crosstalk::filter_slider("cohort_size", "Cohort Size", targets_tcga, ~cohort_size)
  )
)

htmltools::br()
DT::datatable(targets_tcga, escape = F, 
              extensions=c("Buttons","Responsive"), 
              width = "100%",
              options=list(buttons = c('csv','excel'),dom = 'Bfrtip')
)

```


```{r tcga_freqs_missing, echo=F, results = 'asis', eval = missing_tcga_freqs}
cat('<i>No genes with tcga aberrations were found.</i>',sep='\n')
cat('\n')
```

<br><br>
