<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cancer-dedicated gene set interpretation • oncoEnrichR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Cancer-dedicated gene set interpretation">
<meta property="og:description" content="oncoEnrichR performs annotation and prioritization of genesets 
    discovered from high-throughput experiments within cancer biology. The tool 
    queries a number of high-quality resources, including Project Score 
    (cellular fitness in genome-scale drop-out screens), Open Targets Platform 
    (comprehensive target-drug associations, cancer hallmark evidence and 
    target-disease associations), TCGA (genetic aberrations and gene expression 
    patterns in human tumor samples), STRING (protein-protein interaction networks), 
    CORUM (curated protein complexes), COMPPI (subcellular compartments), 
    Human Protein Atlas (tissue and cell-type specific expression patterns and 
    prognostic associations), and Gene Ontology/MSigDB/REACTOME/KEGG/NetPath 
    (functional enrichment/over-representation analysis with respect to gene 
    function and cellular pathways). The package outputs both an interactive HTML 
    report and an Excel workbook in which the user can interrogate the various 
    types of annotations and analyses performed with the query geneset. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">oncoEnrichR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/annotation_sources.html">Gene annotation sources</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sigven/oncoEnrichR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="oncoenrichr---cancer-dedicated-gene-set-interpretation" class="section level2">
<h2 class="hasAnchor">
<a href="#oncoenrichr---cancer-dedicated-gene-set-interpretation" class="anchor"></a>oncoEnrichR - cancer-dedicated gene set interpretation</h2>
<div id="contents" class="section level3">
<h3 class="hasAnchor">
<a href="#contents" class="anchor"></a>Contents</h3>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#news">News</a></li>
<li><a href="#annotation-resources">Annotation resources</a></li>
<li><a href="#example-report">Example report</a></li>
<li><a href="#getting-started">Getting started</a></li>
<li><a href="#contact">Contact</a></li>
</ul>
</div>
<div id="overview" class="section level3">
<h3 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h3>
<p><strong>oncoEnrichR</strong> is an R package for functional interrogation of human genesets in the context of cancer. The package is intended for exploratory analysis and prioritization of a gene list (referred to here as <strong>query set</strong>) from high-throughput cancer biology experiments, e.g. genetic screens (siRNA/CRISPR), protein proximity labeling, or transcriptomics (differential expression). The tool queries a number of high-quality data resources in order to assemble useful gene annotations and analyses in an interactive report (examples from the report shown below).</p>
<p><br></p>
<p><img align="center" border="1," src="reference/figures/oncoenrichr_overview.png"></p>
<p><br><br></p>
<p>The contents of the final report attempts to shed light on the following questions:</p>
<ul>
<li>Which diseases/tumor types are associated with genes in the query set, and to what extent?</li>
<li>Which genes in the query set are attributed with cancer hallmark evidence?</li>
<li>Which proteins in the query sets are druggable in diffferent cancer conditions (early and late clinical development phases)? For other proteins in the query set, what is their likelihood of being druggable?</li>
<li>Which protein complexes are relevant for proteins in the query set?</li>
<li>Which subcellular compartments (nucleus, cytosol, plasma membrane etc) are dominant localizations for proteins in the query set?</li>
<li>Are specific tissues or cell types enriched in the query set, considering tissue/cell-type specific expression patterns of target genes?</li>
<li>Which protein-protein interactions are known within the query set? Are there interactions between members of the query set and other cancer-relevant proteins (e.g. proto-oncogenes, tumor-suppressors or predicted cancer drivers)? Which proteins constitute hubs in the protein-protein interaction network?</li>
<li>Are there specific pathways, biological processes or molecular functions that are enriched within the query set, as compared to a reference/background set?</li>
<li>Which members of the query set are frequently mutated in tumor sample cohorts (TCGA, SNVs/InDels, homozygous deletions, copy number amplifications)?</li>
<li>Which members of the query set are co-expressed (strong negative or positive correlations) with cancer-relevant genes (i.e. proto-oncogenes or tumor suppressors) in tumor sample cohorts (TCGA)?</li>
<li>Which members of the query set are associated with better/worse survival in different cancers, considering high or low gene expression levels, mutation, or copy number status in tumors?</li>
<li>Which members of the query set are associated with cellular loss-of-fitness in CRISPR/Cas9 whole-genome drop out screens of cancer cell lines (i.e. reduction of cell viability elicited by a gene inactivation)? Which targets are prioritized therapeutic targets, considering fitness effects and genomic biomarkers in combination?</li>
</ul>
</div>
<div id="news" class="section level3">
<h3 class="hasAnchor">
<a href="#news" class="anchor"></a>News</h3>
<ul>
<li>June 29th 2021: <strong>1.0.3 release</strong>
<ul>
<li>Minor typo fixes in HTML output</li>
<li>Data upgrade: KEGG, WikiPathways, CancerMine</li>
<li>
<strong>NEW</strong>: web interface to oncoEnrichR: <a href="https://oncotools.elixir.no">oncotools.elixir.no</a>
</li>
</ul>
</li>
<li>June 8th 2021: <strong>1.0.0 release</strong>
<ul>
<li>Added methylation as prognostic feature from <a href="https://www.biorxiv.org/content/10.1101/2021.06.01.446243v1">Smith et al., <em>bioRxiv</em>, 2021</a>
</li>
<li>To keep the size of output Excel/HTML files at managable levels, maximum size of queryset is reduced to <strong>n = 600 genes</strong>
</li>
<li>Fixed assymetrical scales of prognostic Z-score plots</li>
</ul>
</li>
</ul>
</div>
<div id="annotation-resources" class="section level3">
<h3 class="hasAnchor">
<a href="#annotation-resources" class="anchor"></a>Annotation resources</h3>
<p>Data harvested from the following resources form the backbone of <em>oncoEnrichR</em>:</p>
<ul>
<li>
<a href="https://targetvalidation.org">Open Targets Platform</a> - human drug-target associations and comprehensive disease-target associations</li>
<li>
<a href="http://bionlp.bcgsc.ca/cancermine/">CancerMine</a> - literature-mined database of drivers, oncogenes and tumor suppressors in cancer</li>
<li>
<a href="https://portal.gdc.cancer.gov/">The Cancer Genome Atlas</a> - gene aberration frequencies and co-expression patterns in ~9,500 primary tumor samples</li>
<li>
<a href="https://www.proteinatlas.org">The Human Protein Atlas</a> - expression data for healthy human tissues (<a href="https://gtexportal.org/home/">GTex</a>)/cell types, and prognostic gene expression associations in cancer</li>
<li>
<a href="http://software.broadinstitute.org/gsea/msigdb/index.jsp">Molecular Signatures Database (MSigDB)</a> - collection of annotated (e.g. towards pathways) genesets for enrichment/overrepresentation analysis. This includes genesets from <a href="http://geneontology.org/">Gene Ontology</a>, <a href="https://reactome.org/">Reactome</a>, <a href="https://www.genome.jp/kegg/pathway.html">KEGG</a>, <a href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>, <a href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BIOCARTA</a>, as well as curated <a href="https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C7">immunologic</a> and <a href="https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C6">cancer-specific</a> signatures.</li>
<li>
<a href="http://www.netpath.org">NetPath</a> - signaling transduction pathways</li>
<li>
<a href="https://string-db.org">STRING</a> - protein-protein interaction database</li>
<li>
<a href="https://mips.helmholtz-muenchen.de/corum/">CORUM</a> - protein complex database</li>
<li>
<a href="http://comppi.linkgroup.hu/">ComPPI</a> - subcellular compartment database</li>
<li>
<a href="https://score.depmap.sanger.ac.uk">Project Score</a> - Database on the effects on cancer cell line viability elicited by CRISPR-Cas9 mediated gene activation</li>
<li>
<a href="http://survival.cshl.edu/">Genetic determinants of survival in cancer</a> - Resource on the relative prognostic impact of gene mutation, expression, methylation, and CNA in human cancers</li>
</ul>
</div>
<div id="example-report" class="section level3">
<h3 class="hasAnchor">
<a href="#example-report" class="anchor"></a>Example report</h3>
<p><a href="https://doi.org/10.5281/zenodo.4945414"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.4945414.svg" alt="DOI"></a></p>
<p><br></p>
<p><img align="center" width="560" height="750" border="2," src="reference/figures/oncoenrichr_slideshow2.gif"></p>
</div>
<div id="getting-started" class="section level3">
<h3 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h3>
<div id="installation" class="section level4">
<h4 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h4>
<ol>
<li><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages('devtools')</a></code></li>
<li><code><a href="https://devtools.r-lib.org//reference/remote-reexports.html">devtools::install_github('sigven/oncoEnrichR')</a></code></li>
<li><code><a href="https://github.com/sigven/oncoEnrichR">library(oncoEnrichR)</a></code></li>
</ol>
<p><a href="http://github.com/badges/stability-badges"><img src="http://badges.github.io/stability-badges/dist/stable.svg" alt="stable"></a></p>
</div>
<div id="usage" class="section level4">
<h4 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h4>
<p><em>oncoEnrichR</em> performs its operations through the following procedures/methods:</p>
<p><strong>1.</strong> <code><a href="reference/onco_enrich.html">oncoEnrichR::onco_enrich()</a></code></p>
<ul>
<li>
<p>Consists of two main processing steps:</p>
<p>1) Takes an input/query list of human gene/protein identifiers (e.g. UniProt accession, RefSeq/Ensembl transcript identifer etc.) as input and conducts uniform identifier conversion</p>
<p>2) Performs extensive annotation, enrichment and membership analyses of the query set against underlying data sources on cancer-relevant properties of human genes and their interrelationships.</p>
</li>
<li>
<p>Technically, the method returns a <em>list object</em> with all contents of the analyses performed. The specific arguments/options and default values are outlined below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/onco_enrich.html">onco_enrich</a></span><span class="op">(</span>
  <span class="va">query</span>,
  query_id_type <span class="op">=</span> <span class="st">"symbol"</span>,
  ignore_id_err <span class="op">=</span> <span class="cn">TRUE</span>,
  project_title <span class="op">=</span> <span class="st">"Project title"</span>,
  project_owner <span class="op">=</span> <span class="st">"Project owner"</span>,
  project_description <span class="op">=</span> <span class="st">"Project description"</span>,
  bgset <span class="op">=</span> <span class="cn">NULL</span>,
  bgset_id_type <span class="op">=</span> <span class="st">"symbol"</span>,
  bgset_description <span class="op">=</span> <span class="st">"All protein-coding genes"</span>,
  p_value_cutoff_enrichment <span class="op">=</span> <span class="fl">0.05</span>,
  p_value_adjustment_method <span class="op">=</span> <span class="st">"BH"</span>,
  q_value_cutoff_enrichment <span class="op">=</span> <span class="fl">0.2</span>,
  min_geneset_size <span class="op">=</span> <span class="fl">10</span>,
  max_geneset_size <span class="op">=</span> <span class="fl">500</span>,
  min_subcellcomp_confidence <span class="op">=</span> <span class="fl">1</span>,
  subcellcomp_show_cytosol <span class="op">=</span> <span class="cn">FALSE</span>,
  simplify_go <span class="op">=</span> <span class="cn">TRUE</span>,
  ppi_add_nodes <span class="op">=</span> <span class="fl">50</span>,
  ppi_score_threshold <span class="op">=</span> <span class="fl">900</span>,
  show_ppi <span class="op">=</span> <span class="cn">TRUE</span>,
  show_drugs_in_ppi <span class="op">=</span> <span class="cn">TRUE</span>,
  show_disease <span class="op">=</span> <span class="cn">TRUE</span>,
  show_top_diseases_only <span class="op">=</span> <span class="cn">TRUE</span>,
  show_cancer_hallmarks <span class="op">=</span> <span class="cn">TRUE</span>,
  show_drug <span class="op">=</span> <span class="cn">TRUE</span>,
  show_enrichment <span class="op">=</span> <span class="cn">TRUE</span>,
  show_tcga_aberration <span class="op">=</span> <span class="cn">TRUE</span>,
  show_tcga_coexpression <span class="op">=</span> <span class="cn">TRUE</span>,
  show_subcell_comp <span class="op">=</span> <span class="cn">TRUE</span>,
  show_crispr_lof <span class="op">=</span> <span class="cn">TRUE</span>,
  show_cell_tissue <span class="op">=</span> <span class="cn">FALSE</span>,
  show_prognostic_cancer_assoc <span class="op">=</span> <span class="cn">TRUE</span>,
  show_complex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="44%">
<col width="55%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>query</code></td>
<td>character vector with gene/query identifiers</td>
</tr>
<tr class="even">
<td><code>query_id_type</code></td>
<td>character indicating type of identifier used for query set (one of “<em>uniprot_acc</em>”, “<em>symbol</em>”, “<em>entrezgene</em>”, “<em>ensembl_gene</em>”, “<em>refseq_mrna</em>”, “<em>refseq_protein</em>”, “<em>ensembl_protein</em>”, or “<em>ensembl_mrna</em>”)</td>
</tr>
<tr class="odd">
<td><code>ignore_id_err</code></td>
<td>logical indicating if analysis should continue when erroneous/unmatched query identifiers are encountered in query or background gene set</td>
</tr>
<tr class="even">
<td><code>project_title</code></td>
<td>project title (report title)</td>
</tr>
<tr class="odd">
<td><code>project_owner</code></td>
<td>project owner (e.g. lab/PI)</td>
</tr>
<tr class="even">
<td><code>project_description</code></td>
<td>brief description of project, how target list was derived</td>
</tr>
<tr class="odd">
<td><code>bgset</code></td>
<td>character vector with gene identifiers, used as reference/background for enrichment/over-representation analysis</td>
</tr>
<tr class="even">
<td><code>bgset_id_type</code></td>
<td>character indicating type of identifier used for background set (one of “<em>uniprot_acc</em>”, “<em>symbol</em>”, “<em>entrezgene</em>”, “<em>ensembl_gene</em>”, “<em>refseq_mrna</em>”, “<em>refseq_protein</em>”, “<em>ensembl_protein</em>”, or “<em>ensembl_mrna</em>”)</td>
</tr>
<tr class="odd">
<td><code>bgset_description</code></td>
<td>character with description of background gene set (e.g. ‘All lipid-binding proteins (n = 200)’)</td>
</tr>
<tr class="even">
<td><code>p_value_cutoff_enrichment</code></td>
<td>cutoff p-value for enrichment/over-representation analysis</td>
</tr>
<tr class="odd">
<td><code>p_value_adjustment_method</code></td>
<td>one of “holm”, “hochberg”, “hommel”, “bonferroni”, “BH”, “BY”, “fdr”, “none”</td>
</tr>
<tr class="even">
<td><code>q_value_cutoff_enrichment</code></td>
<td>cutoff q-value for enrichment/over-representation analysis</td>
</tr>
<tr class="odd">
<td><code>min_geneset_size</code></td>
<td>minimal size of geneset annotated by term for testing in enrichment/over-representation analysis</td>
</tr>
<tr class="even">
<td><code>max_geneset_size</code></td>
<td>maximal size of geneset annotated by term for testing in enrichment/over-representation analysis</td>
</tr>
<tr class="odd">
<td><code>min_subcellcomp_confidence</code></td>
<td>minimum confidence level of subcellular compartment annotations (range from 1 to 6, 6 is strongest)</td>
</tr>
<tr class="even">
<td><code>subcellcomp_show_cytosol</code></td>
<td>logical indicating if subcellular heatmap should highlight cytosol as a subcellular protein location or not</td>
</tr>
<tr class="odd">
<td><code>simplify_go</code></td>
<td>remove highly similar GO terms in results from GO enrichment/over-representation analysis (recommended)</td>
</tr>
<tr class="even">
<td><code>ppi_add_nodes</code></td>
<td>number of neighbouring nodes to add to query set when computing the protein-protein interaction network (STRING)</td>
</tr>
<tr class="odd">
<td><code>ppi_score_threshold</code></td>
<td>minimum significance score (0-1000) for protein-protein interactions to be included in network (STRING)</td>
</tr>
<tr class="even">
<td><code>show_ppi</code></td>
<td>logical indicating if report should contain protein-protein interaction data of query set and their closely interacting partners (STRING), local network communities, and rank of query set based on network centrality</td>
</tr>
<tr class="odd">
<td><code>show_drugs_in_ppi</code></td>
<td>logical indicating if targeted drugs (&gt;= phase 3) should be displayed in protein-protein interaction network (Open Targets Platform)</td>
</tr>
<tr class="even">
<td><code>show_disease</code></td>
<td>logical indicating if report should contain ranked associations to cancer phenotypes (overall), as well as tumor-type specific rankings (association score &gt;= 0.4, minimum number of sources contributing to association &gt;= 2, (Open Targets Platform))</td>
</tr>
<tr class="odd">
<td><code>show_top_diseases_only</code></td>
<td>logical indicating if report should only show top (20) cancer phenotypes/disease associations from Open Targets Platform</td>
</tr>
<tr class="even">
<td><code>show_cancer_hallmarks</code></td>
<td>logical indicating if report should show evidence that relates query member genes to cancer hallmarks</td>
</tr>
<tr class="odd">
<td><code>show_drug</code></td>
<td>logical indicating if report should contain cancer drugs targeted towards proteins in the query list (early and late development phase) and tractability/druggability data for all query entries, from Open Targets Platform)</td>
</tr>
<tr class="even">
<td><code>show_enrichment</code></td>
<td>logical indicating if report should perform and list functional enrichment/over-representation analysis of query set (MSigDB, GO, KEGG, REACTOME, WikiPathways, NetPath)</td>
</tr>
<tr class="odd">
<td><code>show_tcga_aberration</code></td>
<td>logical indicating if report should contain TCGA aberration plots (amplifications/deletions, SNVs/InDels (oncoplots))</td>
</tr>
<tr class="even">
<td><code>show_tcga_coexpression</code></td>
<td>logical indicating if report should list oncogenes/tumor suppressor genes that significantly correlate with entries in query set in terms of expression (across TCGA cohorts)</td>
</tr>
<tr class="odd">
<td><code>show_subcell_comp</code></td>
<td>logical indicating if report should list subcellular compartment annotations (ComPPI)</td>
</tr>
<tr class="even">
<td><code>show_crispr_lof</code></td>
<td>logical indicating if report should list results from CRISPR/Cas9 loss-of-fitness screens and associated target priority scores (Project Score)</td>
</tr>
<tr class="odd">
<td><code>show_cell_tissue</code></td>
<td>logical indicating if report should list results from tissue (GTex)- and cell-type (HPA) specific gene expression patterns in query set</td>
</tr>
<tr class="even">
<td><code>show_prognostic_cancer_assoc</code></td>
<td>logical indicating if report should list results from significant associations between gene expression, mutation, methylation and survival (Human Protein Atlas - Pathology Atlas, and survival.cshl.edu)</td>
</tr>
<tr class="odd">
<td><code>show_complex</code></td>
<td>logical indicating if report should show membership of target proteins in known protein complexes (CORUM)</td>
</tr>
</tbody>
</table>
</li>
</ul>
<p><strong>2.</strong> <code><a href="reference/write.html">oncoEnrichR::write()</a></code></p>
<ul>
<li>
<p>Consists of two main processing steps:</p>
<p>1) Transform the contents of the analyses returned by <em>oncoEnrichR::onco_enrich()</em> into various visualizations and interactive tables</p>
<p>2) Assemble and write the final analysis report through</p>
<ul>
<li>A) a structured and interactive <em>oncoEnrichR</em> HTML report, OR</li>
<li>B) a multisheet Excel workbook with results from the annotations and analyses provided by <em>oncoEnrichR</em>
</li>
</ul>
</li>
</ul>
</div>
<div id="example-run" class="section level4">
<h4 class="hasAnchor">
<a href="#example-run" class="anchor"></a>Example run</h4>
<p>A target list of <em>n = 134</em> high-confidence interacting proteins with the c-MYC oncoprotein were previously identified through BioID protein proximity assay in standard cell culture and in tumor xenografts (<a href="https://www.ncbi.nlm.nih.gov/pubmed/25452129">Dingar et al., J Proteomics, 2015</a>). We ran this target list through the <em>oncoEnrichR</em> analysis workflow using the following configurations for the <code>onco_enrich</code> method:</p>
<ul>
<li><code>project_title = "cMYC_BioID_screen"</code></li>
<li><code>project_owner = "Raught et al."</code></li>
</ul>
<p>and produced the <a href="https://doi.org/10.5281/zenodo.4945414">following HTML report with results</a>.</p>
<p>Below are R commands provided to reproduce the example output. <strong>NOTE</strong>: Replace “LOCAL_FOLDER” with a directory on your local computer:</p>
<ul>
<li><code><a href="https://github.com/sigven/oncoEnrichR">library(oncoEnrichR)</a></code></li>
<li><code>myc_interact_targets &lt;- read.csv(system.file("extdata","myc_data.csv", package = "oncoEnrichR"), stringsAsFactors = F)</code></li>
<li><code>myc_report &lt;- oncoEnrichR::onco_enrich(query = myc_interact_targets$symbol, show_cell_tissue = T, project_title = "cMYC_BioID_screen", project_owner = "Raught et al.")</code></li>
<li><code><a href="reference/write.html">oncoEnrichR::write(report = myc_report, file = "LOCAL_FOLDER/myc_report_oncoenrichr.html", format = "html")</a></code></li>
<li><code><a href="reference/write.html">oncoEnrichR::write(report = myc_report, file = "LOCAL_FOLDER/myc_report_oncoenrichr.xlsx", format = "excel")</a></code></li>
</ul>
</div>
</div>
<div id="contact" class="section level3">
<h3 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h3>
<p>sigven AT ifi.uio.no</p>
</div>
<div id="funding--collaboration" class="section level3">
<h3 class="hasAnchor">
<a href="#funding--collaboration" class="anchor"></a>Funding &amp; collaboration</h3>
<p>OncoEnrichR is supported by the <a href="https://www.med.uio.no/cancell/english/">Centre for Cancer Cell Reprogramming</a> at the <a href="https://www.uio.no">University of Oslo</a>/<a href="https://radium.no">Oslo University Hospital</a>, and <a href="https://elixir.no/organization/organisation/elixir-uio">Elixir Norway (Oslo node)</a>.</p>
<p><br><br></p>
<p float="left">
<a href="https://www.med.uio.no/cancell/english/"> <img src="reference/figures/can-cell.png" width="150"></a>         <a href="https://elixir.no/organization/organisation/elixir-uio"> <img src="reference/figures/elixir_norway.png" width="200"></a>
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/sigven/oncoEnrichR/">https://​github.com/​sigven/​oncoEnrichR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/sigven/oncoEnrichR/issues">https://​github.com/​sigven/​oncoEnrichR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/mit-license.php">MIT</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Sigve Nakken <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8468-2050" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Sigve Nakken.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
