<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Interrogate a gene list for cancer relevance — onco_enrich • oncoEnrichR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Interrogate a gene list for cancer relevance — onco_enrich"><meta property="og:description" content="Interrogate a gene list for cancer relevance"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">oncoEnrichR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/annotation_sources.html">Gene annotation sources</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/sigven/oncoEnrichR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Interrogate a gene list for cancer relevance</h1>
    <small class="dont-index">Source: <a href="https://github.com/sigven/oncoEnrichR/blob/HEAD/R/onco_enrichr.R" class="external-link"><code>R/onco_enrichr.R</code></a></small>
    <div class="hidden name"><code>onco_enrich.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Interrogate a gene list for cancer relevance</p>
    </div>

    <div class="ref-usage sourceCode"><pre class="sourceCode r"><code><span class="fu">onco_enrich</span><span class="op">(</span>
  query <span class="op">=</span> <span class="cn">NULL</span>,
  oeDB <span class="op">=</span> <span class="cn">NULL</span>,
  query_id_type <span class="op">=</span> <span class="st">"symbol"</span>,
  ignore_id_err <span class="op">=</span> <span class="cn">TRUE</span>,
  html_floating_toc <span class="op">=</span> <span class="cn">T</span>,
  html_report_theme <span class="op">=</span> <span class="st">"default"</span>,
  project_title <span class="op">=</span> <span class="st">"Project title"</span>,
  project_owner <span class="op">=</span> <span class="st">"Project owner"</span>,
  project_description <span class="op">=</span> <span class="st">"Project description"</span>,
  bgset <span class="op">=</span> <span class="cn">NULL</span>,
  bgset_id_type <span class="op">=</span> <span class="st">"symbol"</span>,
  bgset_description <span class="op">=</span> <span class="st">"All protein-coding genes"</span>,
  p_value_cutoff_enrichment <span class="op">=</span> <span class="fl">0.05</span>,
  p_value_adjustment_method <span class="op">=</span> <span class="st">"BH"</span>,
  q_value_cutoff_enrichment <span class="op">=</span> <span class="fl">0.2</span>,
  min_geneset_size <span class="op">=</span> <span class="fl">10</span>,
  max_geneset_size <span class="op">=</span> <span class="fl">500</span>,
  num_terms_enrichment_plot <span class="op">=</span> <span class="fl">20</span>,
  min_subcellcomp_confidence <span class="op">=</span> <span class="fl">1</span>,
  subcellcomp_show_cytosol <span class="op">=</span> <span class="cn">FALSE</span>,
  min_confidence_reg_interaction <span class="op">=</span> <span class="st">"D"</span>,
  simplify_go <span class="op">=</span> <span class="cn">TRUE</span>,
  ppi_add_nodes <span class="op">=</span> <span class="fl">50</span>,
  ppi_score_threshold <span class="op">=</span> <span class="fl">900</span>,
  show_ppi <span class="op">=</span> <span class="cn">TRUE</span>,
  show_drugs_in_ppi <span class="op">=</span> <span class="cn">TRUE</span>,
  show_disease <span class="op">=</span> <span class="cn">TRUE</span>,
  show_top_diseases_only <span class="op">=</span> <span class="cn">TRUE</span>,
  show_cancer_hallmarks <span class="op">=</span> <span class="cn">TRUE</span>,
  show_drug <span class="op">=</span> <span class="cn">TRUE</span>,
  show_enrichment <span class="op">=</span> <span class="cn">TRUE</span>,
  show_tcga_aberration <span class="op">=</span> <span class="cn">TRUE</span>,
  show_tcga_coexpression <span class="op">=</span> <span class="cn">TRUE</span>,
  show_cell_tissue <span class="op">=</span> <span class="cn">FALSE</span>,
  show_ligand_receptor <span class="op">=</span> <span class="cn">TRUE</span>,
  show_regulatory_interactions <span class="op">=</span> <span class="cn">TRUE</span>,
  show_unknown_function <span class="op">=</span> <span class="cn">TRUE</span>,
  show_prognostic_cancer_assoc <span class="op">=</span> <span class="cn">TRUE</span>,
  show_subcell_comp <span class="op">=</span> <span class="cn">TRUE</span>,
  show_crispr_lof <span class="op">=</span> <span class="cn">TRUE</span>,
  show_complex <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>

    <div id="ref-sections">
    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>query</dt>
<dd><p>character vector with gene/query identifiers</p></dd>
<dt>oeDB</dt>
<dd><p>oncoEnrichR annotation database - as returned from oncoEnrichR::load_db()</p></dd>
<dt>query_id_type</dt>
<dd><p>character indicating source of query (one of "uniprot_acc", "symbol",
"entrezgene", or "ensembl_gene","ensembl_mrna","refseq_mrna","ensembl_protein","refseq_protein")</p></dd>
<dt>ignore_id_err</dt>
<dd><p>logical indicating if analysis should continue when uknown query identifiers are encountered</p></dd>
<dt>html_floating_toc</dt>
<dd><p>logical - float the table of contents to the left of the main document content (HTML report). The floating table of contents will always be visible even when the document is scrolled</p></dd>
<dt>html_report_theme</dt>
<dd><p>Bootswatch theme for HTML report (any of "bootstrap","cerulean","cosmo","default","flatly","journal","lumen","paper","sandstone","simplex","spacelab","united","yeti")</p></dd>
<dt>project_title</dt>
<dd><p>project title (title of report)</p></dd>
<dt>project_owner</dt>
<dd><p>name of project owner</p></dd>
<dt>project_description</dt>
<dd><p>project background information</p></dd>
<dt>bgset</dt>
<dd><p>character vector with gene identifiers, used as reference/background for enrichment/over-representation analysis</p></dd>
<dt>bgset_id_type</dt>
<dd><p>character indicating source of background ("uniprot_acc","symbol","entrezgene","ensembl_gene_id")</p></dd>
<dt>bgset_description</dt>
<dd><p>character indicating type of background (e.g. "All lipid-binding proteins (n = 200)")</p></dd>
<dt>p_value_cutoff_enrichment</dt>
<dd><p>cutoff p-value for enrichment/over-representation analysis</p></dd>
<dt>p_value_adjustment_method</dt>
<dd><p>one of "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"</p></dd>
<dt>q_value_cutoff_enrichment</dt>
<dd><p>cutoff q-value for enrichment analysis</p></dd>
<dt>min_geneset_size</dt>
<dd><p>minimal size of geneset annotated by term for testing in enrichment/over-representation analysis</p></dd>
<dt>max_geneset_size</dt>
<dd><p>maximal size of geneset annotated by term for testing in enrichment/over-representation analysis</p></dd>
<dt>num_terms_enrichment_plot</dt>
<dd><p>number of top enriched Gene Ontology terms (max) to show in enrichment barplot</p></dd>
<dt>min_subcellcomp_confidence</dt>
<dd><p>minimum confidence level of subcellular compartment annotations (range from 1 to 6, 6 is strongest)</p></dd>
<dt>subcellcomp_show_cytosol</dt>
<dd><p>logical indicating if subcellular heatmap should show highlight proteins located in the cytosol or not</p></dd>
<dt>min_confidence_reg_interaction</dt>
<dd><p>minimum confidence level for regulatory interactions (TF-target) retrieved from DoRothEA ('A','B','C', or 'D')</p></dd>
<dt>simplify_go</dt>
<dd><p>remove highly similar GO terms in results from GO enrichment/over-representation analysis</p></dd>
<dt>ppi_add_nodes</dt>
<dd><p>number of nodes to add to target set when computing the protein-protein interaction network (STRING)</p></dd>
<dt>ppi_score_threshold</dt>
<dd><p>minimum score (0-1000) for retrieval of protein-protein interactions (STRING)</p></dd>
<dt>show_ppi</dt>
<dd><p>logical indicating if report should contain protein-protein interaction data (STRING)</p></dd>
<dt>show_drugs_in_ppi</dt>
<dd><p>logical indicating if targeted drugs (&gt; phase 3) should be displayed in protein-protein interaction network (Open Targets Platform)</p></dd>
<dt>show_disease</dt>
<dd><p>logical indicating if report should contain disease associations (Open Targets Platform, association_score &gt;= 0.05, support from at least two data types)</p></dd>
<dt>show_top_diseases_only</dt>
<dd><p>logical indicating if report should contain top (n = 20) disease associations only pr. query gene (Open Targets Platform)</p></dd>
<dt>show_cancer_hallmarks</dt>
<dd><p>logical indicating if report should contain annotations/evidence of cancer hallmarks per query gene (COSMIC/Open Targets Platform)</p></dd>
<dt>show_drug</dt>
<dd><p>logical indicating if report should contain targeted cancer drug information</p></dd>
<dt>show_enrichment</dt>
<dd><p>logical indicating if report should contain functional enrichment/over-representation analysis (MSigDB, GO, KEGG, REACTOME, NetPath, WikiPathways)</p></dd>
<dt>show_tcga_aberration</dt>
<dd><p>logical indicating if report should contain TCGA aberration plots (amplifications/deletions)</p></dd>
<dt>show_tcga_coexpression</dt>
<dd><p>logical indicating if report should contain TCGA co-expression data (RNAseq) of query set with oncogenes/tumor suppressor genes</p></dd>
<dt>show_cell_tissue</dt>
<dd><p>logical indicating if report should contain tissue-specificity and single cell-type specificity assessments (Human Protein Atlas)
of target genes, using data from the Human Protein Atlas</p></dd>
<dt>show_ligand_receptor</dt>
<dd><p>logical indicating if report should contain ligand-receptor interactions (CellChatDB)</p></dd>
<dt>show_regulatory_interactions</dt>
<dd><p>logical indicating if report should contain data on transcription factor (TF) - target interactions relevant for the query set (DoRothEA)</p></dd>
<dt>show_unknown_function</dt>
<dd><p>logical indicating if report should highlight target genes with unknown or poorly defined functions (GO/Uniprot KB/NCBI)</p></dd>
<dt>show_prognostic_cancer_assoc</dt>
<dd><p>logical indicating if mRNA-based (single-gene) prognostic associations to cancer types should be listed (Human Protein Atlas/TCGA)</p></dd>
<dt>show_subcell_comp</dt>
<dd><p>logical indicating if report should provide subcellular compartment annotations (ComPPI)</p></dd>
<dt>show_crispr_lof</dt>
<dd><p>logical indicating if report should provide fitness scores and target priority scores from CRISPR/Cas9 loss-of-fitness screens (Project Score)</p></dd>
<dt>show_complex</dt>
<dd><p>logical indicating if report should provide target memberships in known protein complexes (ComplexPortal/Compleat/PDB/CORUM)</p></dd>
<dt>...</dt>
<dd><p>arguments for Galaxy/web-based processing</p></dd>
</dl></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Sigve Nakken.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

